server/service/src/document/document_service.rs

```diff
@@ -0,0 +1,440 @@
+    common_ancestor::{common_ancestors, AncestorDB, CommonAncestorError, InMemoryAncestorDB},
+    merge::{three_way_merge, two_way_merge, TakeLatestConflictSolver},
+    raw_document::RawDocument,
+    topological_sort::{extract_tree, topo_sort},
+};
+
+#[derive(Debug)]
+pub enum DocumentInsertError {
+    /// Input document doesn't match the provided json schema
+    InvalidDataSchema(Vec<String>),
+    /// Document version needs to be merged first. Contains an automerged document which can be
+    /// reviewed and/or inserted.
+    /// If no automerged document is provided the document couldn't be automerged, e.g because the
+    /// merged data has an invalid data schema.
+    MergeRequired(Option<RawDocument>),
+    InvalidDocumentHistory,
+    /// Unable to finalise a document (assign an id)
+    FinalisationError(String),
+    DatabaseError(RepositoryError),
+    InternalError(String),
+}
+
+#[derive(Debug)]
+pub enum DocumentHistoryError {
+    DatabaseError(RepositoryError),
+    GraphError(String),
+}
+
+impl From<RepositoryError> for DocumentHistoryError {
+    fn from(err: RepositoryError) -> Self {
+        DocumentHistoryError::DatabaseError(err)
+    }
+}
+
+pub trait DocumentServiceTrait: Sync + Send {
+    fn get_document(
+        &self,
+        ctx: &ServiceContext,
+        store_id: &str,
+        name: &str,
+    ) -> Result<Option<Document>, RepositoryError> {
+        DocumentRepository::new(&ctx.connection).find_one_by_name(store_id, name)
+    }
+
+    fn get_documents(
+        &self,
+        ctx: &ServiceContext,
+        store_id: &str,
+        filter: Option<DocumentFilter>,
+    ) -> Result<Vec<Document>, RepositoryError> {
+        let filter = filter.map(|mut f| {
+            f.store_id = Some(EqualFilter::equal_to(store_id));
+            f
+        });
+        DocumentRepository::new(&ctx.connection).query(filter)
+    }
+
+    fn get_document_history(
+        &self,
+        ctx: &ServiceContext,
+        store_id: &str,
+        name: &str,
+    ) -> Result<Vec<Document>, DocumentHistoryError> {
+        let repo = DocumentRepository::new(&ctx.connection);
+        let head = match repo.head(store_id, name)? {
+            Some(head) => head,
+            None => return Ok(vec![]),
+        };
+
+        let docs = repo.document_history(name)?;
+
+        // We might have Documents from different stores in our repo; extract our tree:
+        let graph =
+            extract_tree(head.head, docs).map_err(|err| DocumentHistoryError::GraphError(err))?;
+        let sorted = topo_sort(graph).map_err(|err| DocumentHistoryError::GraphError(err))?;
+        Ok(sorted)
+    }
+
+    fn update_document(
+        &self,
+        ctx: &ServiceContext,
+        store_id: &str,
+        doc: RawDocument,
+    ) -> Result<Document, DocumentInsertError> {
+        let document = ctx
+            .connection
+            .transaction_sync(|con| {
+                let validator = json_validator(con, &doc)?;
```

Comment:
I think validate method is clearer, you've previously pointed out that it's hard to see what errors are returned/mapped. With validate method it would be:
```rust
/* validate */

let schema = /* try get schema with repo */.map_err( /* database err */)?.ok_or(/* schema does not exist */)?
let validator = /*  compile */.map_err(/* internal error, although would prefer concrete err and graphql can map to internal */)?
/* validate json */.map_err(/* InvalidDataSchema */)?
```

Reply:
The internal errors are really just internal errors, it shouldn't be allowed in the first place to insert an invalid schema...